File: leetcode_python/33. Search in Rotated Sorted Array.py
class Solution:
    def search(self, nums: [int], target: int) -> int:
        if not nums: return -1
        i, j = 0, len(nums) - 1
        while i < j:
            m = (i + j) // 2
            if nums[m] < nums[j]: j = m
            else: i = m + 1
        if nums[i] <= target <= nums[-1]: i, j = i, len(nums) - 1
        elif nums[0] <= target <= nums[i - 1]: i, j = 0, i - 1
        else: return -1
        while i <= j:
            m = (i + j) // 2
            if nums[m] > target: j = m - 1
            elif nums[m] < target: i = m + 1
            else: return m
        return -1

File: leetcode_java/33. Search in Rotated Sorted Array.java
class Solution {
    public int search(int[] nums, int target) {
        if(nums.length == 0) return -1;
        int i = 0, j = nums.length - 1;
        while(i < j){
            int m = (i + j) / 2;
            if(nums[m] < nums[j]) j = m;
            else i = m + 1;
        }
        if(i == 0 || nums[i] <= target && target <= nums[nums.length - 1])
            j = nums.length - 1;
        else if(nums[0] <= target && target <= nums[i - 1]){
            j = i - 1;
            i = 0;
        }
        else return -1;
        while(i <= j) {
            int m = (i + j) / 2;
            if(nums[m] < target) i = m + 1;
            else if(nums[m] > target) j = m - 1;
            else return m;
        }
        return -1;
    }
}